### PR TITLE
Add Microsoft Visio stencil for icons

### PR DESCRIPTION
**Pull request by dahovey**
Hopefully this is of use to others. Request includes Microsoft Visio Stencil file, with all PNG files within [./icons/png/](./icons/png/) that are labeled. Shapes have been sized so that the original 128x128 images are used, but the shape itself is resized to be 32x32.

My use case, for this addition, is to model cloud infrastructure along with Kubernetes resources.

### CODE REVIEW

**Comment by dahovey** [comment]
I signed it.

**Comment by dahovey** [comment]
Well, thanks for not shooting down right away. 
Interesting challenge accepted.

**Comment by dahovey** [comment]
Ok, here are my thoughts and ultimately my attempt to convince you. It is possible to programmatically generate a Visio file (using OCX/Interop), but it would require Visio to be installed, which sounds like it isn't an option.

I found this documentation from Microsoft [https://docs.microsoft.com/en-us/office/client-developer/visio/how-to-manipulate-the-visio-file-format-programmatically](https://docs.microsoft.com/en-us/office/client-developer/visio/how-to-manipulate-the-visio-file-format-programmatically). This would only require .NET Framework and so, in theory, could be hosted from a Windows-based dockerfile.

Proof-of-concept repository created at [https://github.com/dahovey/visio-stencil-creator](https://github.com/dahovey/visio-stencil-creator). `Program.cs` is where the magic happens. The code needs lots of cleanup, but just illustrates it can be done. Small app looks for any files with png extension in specific folder. The app uses a template visio file, and modifies it to include new visio shapes based on png files.

Provided it was cleaned up, would this be acceptable solution?

**Comment by dahovey** [comment]
Updated repository [https://github.com/dahovey/visio-stencil-creator](https://github.com/dahovey/visio-stencil-creator). Changed dependency to Microsoft .NET Core (with linux Docker image). Pushed image to `hoveytech/visio-stencil-creator`.

Docker command used: 

```shell
docker run -v C:\HoveyTech\kubernetes.community\icons:/app/out -v C:\HoveyTech\kubernetes.community\icons\png:/app/test-data hoveytech/visio-stencil-creator:v0.1 "resources/labeled/*.png;infrastructure_components/labeled/*.png;control_plane_components/labeled/*.png" "/app/test-data" "/app/out/kubernetes-visio-stencil.vssx"
```
`C:\HoveyTech\kubernetes.community` is my cloned version of this repo.

Above can be used in automated builds to update stencil file.

**Comment by thockin** [comment]
Thanks for rising to the challenge.  As long as this tool is properly
cross-platform and deterministic, I don't see why we can't take it.

You'll need to package it up under tools/ and get someone to sign up
to review that .net code

**Comment by dahovey** [comment]
for clarification:

* add in tools directory...bash script that depends on image in docker hub? Can be the image already created, or another tag/image? Script could run container, referencing paths to png images within `icons` folder.
* add in tools directory...bash script all C# code and Dockerfile so no outside dependencies?

**Which is preferred for your use?**
**How do I `get someone to sign up to review that .net code`?**

I like the docker image idea in docker hub, based on the concept it could be used by other projects.

**Comment by thockin** [comment]
I don't think you need to stick all the C# code in there.  It would be nice
to have a way to rebuild the docker image locally, rather than depend on an
arbitrary 3rd-party binary tool, but I'll consider that "extra credit".

Basically, as a contributor, if I add a new icon I should be able to run
./icons/tools/update-everything.sh and have the PNGs rendered, the visio
accumulated, etc.

**Comment by dahovey** [comment]
Ok, finally, here it is.

- Added `update-everything.sh` and `update-visio.sh` script in tools folder.
- Based on usage of `rasterize.sh`, all scripts within `tools` path are expected to be run from `icons` folder. (i.e. `./tools/update-everything.sh`).
- Script `update-visio.sh` builds .NET Core project from source in external repository and places generated Visio Stencil in `visio` folder.

Let me know any comments. Thanks!

**Comment by dahovey** [comment]
Are the 128x128 images going to be removed in favor 256x256? It appears the `raterize.sh` script is set for 256?

**Comment by dahovey** [comment]
Originally I created script in `hoveytechllc/visio-stencil-creator` repository that would clone repository and build docker image, but instead of that I put the script in this repository. Logic is a little cleaner with new version.

As well the image pattern for generating the visio stencil, only uses `256.png` files and the `labeled` versions.

Thanks.

File: icons/tools/update-visio.sh [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+GITHUB_ORG="hoveytechllc"
+REPO_NAME="visio-stencil-creator"
+
+rm -fdr ./tools/${REPO_NAME}
+
+# Clone repository in current path
+git clone https://github.com/${GITHUB_ORG}/${REPO_NAME}.git ./tools/${REPO_NAME}
+
+# build image using Dockerfile from github repository
+docker build \
```
Comment (thockin) [inline icons/tools/update-visio.sh]: What if, rather than git clone & build, we pushed the resulting image to k8s.gcr.io and just run from that name?

Comment (dahovey) [reply icons/tools/update-visio.sh]: Sure. Either way is fine to me. We can remove `git clone` & `build` all together, and depend on a pre-built image, or modify to `git clone` & `build/push/run` using `k8s.gcr.io/...` tag.

**Comment by thockin** [comment]
If I build and push the image to GCR this week, you can adapt the
scripts here to use it.  We still need the build script, but we can
document that as "someone builds and pushes it, most users never need to do
that, just consume it", right?

**Comment by dahovey** [comment]
Makes sense. `update-everything.sh` can rasterize svgs, and consume the pre-built `k8s.gcr.io` image for Visio. 

I can make a separate script that will git clone & build/push the image. This separate script can be run manually if anything changes. Let me know what tag you want to use.

**Comment by thockin** [comment]
tag it as v1.0.0 or similar (and tag the code in github to match?)

**Comment by thockin** [comment]
AFAICT the ball is in your court - if you tag the github code I will run a build and push to GCR.

Ideally the Dockerfile and build script would be part of the source repo for the tool, so this PR becomes very simple.  Also, please include the source (the git dir and license) in the docker image

**Comment by dahovey** [comment]
Ok, finally created new script `build-visio-stencil-creator.sh` in the tools folder. It should only need to be run if fixes or changes are made to the code that actually does the stencil generation. It pushes to repository `k8s.gcr.io/visio-stencil-creator:v1.0`

The source code repository is tagged to match (i.e. `release/v1.0`). So the script pulls from a tag `release/<tag variable>` and tags the associated Docker image `gcr.k8s.io/visio-stencil-creator:<tag variable>`. Hope that makes sense.

Then the `update-visio.sh` script just depends on Docker image in GCR.

**Comment by jdumars** [comment]
Not sure if Tim has notifications on or not, but @thockin any chance we can unblock this?

File: icons/tools/update-everything.sh [hunk 2]
```diff
@@ -0,0 +1,4 @@
+#/bin/bash
+
+./tools/rasterize.sh
+./tools/update-visio.sh
```
Comment (thockin) [inline icons/tools/update-everything.sh]: It seems your editor doesn't add newlines - almost every file has this warnign

File: icons/tools/build-visio-stencil-creator.sh [hunk 3]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
```
Comment (thockin) [inline icons/tools/build-visio-stencil-creator.sh]: I checked out the PR branch, and this file is not executable.

File: icons/tools/build-visio-stencil-creator.sh [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+GITHUB_ORG="hoveytechllc"
+REPO_NAME="visio-stencil-creator"
+TAG_NAME="v1.0"
+
+GCR_REPO="gcr.k8s.io/${REPO_NAME}:${TAG_NAME}"
```
Comment (thockin) [inline icons/tools/build-visio-stencil-creator.sh]: This has to be `staging-k8s.gcr.io/...` so I can push it.  promoter moves it to final location.

**Comment by thockin** [comment]
Thanks!

/lgtm
/approve

**Comment by thockin** [comment]
@dahovey It seems that the .vssx file is different every time I generate it. 

Running a hexdump-and-diff shows me a lot of deltas.  Some are obvious - there's are string UUIDs that are changing.  Most are not.  Can you take a look and see if we can do something to make back-to-back runs not produce a diff?

**Pull request merged**